nntrainer/tensor/memory_pool.cpp

```diff
@@ -0,0 +1,258 @@
+ * any allocation but rather just plans the layout and stores the layout.
+ * Subsequent call to this function will overwrite any existing layout.
+ */
+double MemoryPool::planLayout(const MemoryPlanner &planner) {
+  if (mem_pool != nullptr)
+    /** mem_pool must be deallocated when planLayout is being called */
+    throw std::runtime_error("Planning memory layout after allocation");
+
+  if (memory_size.empty())
+    throw std::runtime_error("Planning memory layout for empty pool");
+
+  /** calculate min_pool_size if not already calculated */
+  if (min_pool_size == 0)
+    min_pool_size = calcMinMemoryRequirement();
+
+  pool_size = planner.planLayout(memory_size, memory_validity, memory_offset);
+  if (pool_size < min_pool_size || !validateLayout())
+    throw std::runtime_error("Planned layout is not feasible");
+
+  return double(pool_size) / double(min_pool_size);
+}
+
+/**
+ * @brief Do the allocation of memory
+ *
+ */
+void MemoryPool::allocate() {
+  if (pool_size == 0)
+    throw std::runtime_error("Allocating memory pool with size 0");
+
+  mem_pool = malloc(pool_size);
+}
+
+/**
+ * @brief Get the allocated memory
+ *
+ */
+void *MemoryPool::getMemory(unsigned int idx) {
+  if (mem_pool == nullptr)
+    throw std::invalid_argument("Getting memory before allocation");
+
+  return static_cast<char *>(mem_pool) + memory_offset.at(idx);
+}
+
+/**
+ * @brief Free all the allocated memory
+ *
+ */
+void MemoryPool::deallocate() {
+  if (mem_pool != nullptr)
+    free(mem_pool);
+}
+
+/**
+ * @brief Get the maximum real memory requirement
+ *
+ */
+size_t MemoryPool::size() { return pool_size; }
+
+/**
+ * @brief Get the minimum theoretical memory requirement
+ *
+ */
+size_t MemoryPool::minMemoryRequirement() {
+  if (memory_size.size() && min_pool_size == 0)
+    min_pool_size = calcMinMemoryRequirement();
+
+  return min_pool_size;
+}
+
+/**
+ * @brief Validate the provided layout so that no two memories to be used at
+ * overlap interval has overlapping memories
+ */
+bool MemoryPool::validateLayout() {
+  if (memory_offset.size() != memory_size.size())
+    return false;
+
+  if (memory_size.empty())
+    return pool_size == 0;
+
+  return validateOverflow() && validateOverlap();
+}
+
+/**
+ * @brief Validate the provided layout does not overflow outside the given
+ * size of the memory pool
+ */
+bool MemoryPool::validateOverflow() {
+  for (unsigned int idx = 0; idx < memory_size.size(); idx++)
+    if (memory_offset[idx] + memory_size[idx] > pool_size)
+      return false;
+
+  return true;
+}
+
+/**
+ * @brief check if the two given intervals overlap
+ *
+ * @param s1 start of interval 1
+ * @param e1 end of interval 1
+ * @param s2 start of interval 2
+ * @param e2 end of interval 2
+ *
+ * @return true if overlap else false
+ *
+ * @note overlap check assumes are start is inclusive and end is exclusive
+ */
+template <typename T> static bool overlap(T s1, T e1, T s2, T e2) {
+  if (e1 <= s2 || e2 <= s1)
```

Comment:
1. Super minor, maybe return condition itself?
2. if this is not in critical path, how about adding a check to throw if s1 < e1 and s2 < e2?

Reply:
1. updated.
2. added under debug